cdap-hbase-compat-base/src/main/java/co/cask/cdap/replication/TableUpdater.java

```diff
@@ -91,9 +91,16 @@ public void run() {
             createTableIfNotExists(conf);
             tableExists = true;
           }
-          if (!cachedUpdates.isEmpty()) {
-            LOG.debug("Update Replication State table now. {} entries.", cachedUpdates.size());
-            writeState(cachedUpdates);
+          HashMap<String, Long> cachedUpdatesCopy;
+          synchronized (cachedUpdates) {
+            // Make a copy so HBase write can happen outside of the synchronized block
+            cachedUpdatesCopy = (HashMap<String, Long>) cachedUpdates.clone();
```

Comment:
you can synchronize on any object you want. The concern here is that you'll block update calls until you copy the map.